cluster-autoscaler/observers/scale_state_observer.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package observers
+
+import (
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/klog/v2"
+)
+
+// NGChangeObserver is an observer of:
+// * scale-up(s) for a nodegroup
+// * scale-down(s) for a nodegroup
+// * scale-up failure(s) for a nodegroup
+// * scale-down failure(s) for a nodegroup
+type NGChangeObserver interface {
+	// RegisterScaleUp records scale up for a nodegroup.
+	RegisterScaleUp(nodeGroup cloudprovider.NodeGroup, delta int, currentTime time.Time) error
+	// RegisterScaleDowns records scale down for a nodegroup.
+	RegisterScaleDown(nodeGroup cloudprovider.NodeGroup, nodeName string, currentTime time.Time, expectedDeleteTime time.Time) error
+	// RegisterFailedScaleUp records failed scale-up for a nodegroup.
+	RegisterFailedScaleUp(nodeGroup cloudprovider.NodeGroup, reason string, gpuResourceName, gpuType string, currentTime time.Time) error
+	// RegisterFailedScaleDown records failed scale-down for a nodegroup.
+	RegisterFailedScaleDown(nodeGroup cloudprovider.NodeGroup, reason string, currentTime time.Time) error
```

Comment:
I have removed the code to return errors for all the functions except `RegisterScaleUp` because I want to use it as a replacement for `RegisterOrUpdateScaleUp` and return an error in case a negative delta is passed to `RegisterScaleUp` . More info in https://github.com/kubernetes/autoscaler/pull/5729#discussion_r1427101591

Open to any other options/ideas.

Reply:
Checking if an int is negative is simple enough to be done before making a call to `RegisterScaleUp`. It would then be possible to remove the `error` return type.